File: coding-interview/python/p05_binary_tree.py
"""面试手撕代码（05）：二叉树 —— 配套解法与测试。

https://arganzheng.life/coding-interview-binary-tree.html
"""
from __future__ import annotations

import unittest
from collections import defaultdict, deque
from typing import List, Optional


class TreeNode:
    def __init__(self, val: int = 0, left: "Optional[TreeNode]" = None, right: "Optional[TreeNode]" = None):
        self.val, self.left, self.right = val, left, right


def build(values: List[Optional[int]]) -> Optional[TreeNode]:
    """LeetCode 风格的层序数组（None 为空）建树。"""
    if not values or values[0] is None:
        return None
    root = TreeNode(values[0])
    q = deque([root])
    i = 1
    while q and i < len(values):
        node = q.popleft()
        if i < len(values) and values[i] is not None:
            node.left = TreeNode(values[i]); q.append(node.left)
        i += 1
        if i < len(values) and values[i] is not None:
            node.right = TreeNode(values[i]); q.append(node.right)
        i += 1
    return root


def inorder_iterative(root: Optional[TreeNode]) -> List[int]:
    """LC 94. 迭代中序：一路向左压栈，弹出即访问，再转右。O(n) / O(h)。"""
    out, stack, cur = [], [], root
    while cur or stack:
        while cur:
            stack.append(cur)
            cur = cur.left
        cur = stack.pop()
        out.append(cur.val)
        cur = cur.right
    return out


def preorder_iterative(root: Optional[TreeNode]) -> List[int]:
    """LC 144. 栈：先压右再压左。"""
    out, stack = [], [root] if root else []
    while stack:
        node = stack.pop()
        out.append(node.val)
        if node.right:
            stack.append(node.right)
        if node.left:
            stack.append(node.left)
    return out


def level_order(root: Optional[TreeNode]) -> List[List[int]]:
    """LC 102. BFS 按层：每轮先记下队列长度。O(n) / O(w)。"""
    out: List[List[int]] = []
    q = deque([root] if root else [])
    while q:
        level = []
        for _ in range(len(q)):
            node = q.popleft()
            level.append(node.val)
            if node.left:
                q.append(node.left)
            if node.right:
                q.append(node.right)
        out.append(level)
    return out


def right_side_view(root: Optional[TreeNode]) -> List[int]:
    """LC 199. 每层最后一个。"""
    return [level[-1] for level in level_order(root)]


def max_depth(root: Optional[TreeNode]) -> int:
    """LC 104. 后序：1 + max(左, 右)。"""
    return 0 if not root else 1 + max(max_depth(root.left), max_depth(root.right))


def invert_tree(root: Optional[TreeNode]) -> Optional[TreeNode]:
    """LC 226."""
    if root:
        root.left, root.right = invert_tree(root.right), invert_tree(root.left)
    return root


def is_symmetric(root: Optional[TreeNode]) -> bool:
    """LC 101. 两棵子树镜像比较。"""
    def mirror(a, b):
        if not a and not b:
            return True
        if not a or not b or a.val != b.val:
            return False
        return mirror(a.left, b.right) and mirror(a.right, b.left)
    return mirror(root.left, root.right) if root else True


def diameter_of_binary_tree(root: Optional[TreeNode]) -> int:
    """LC 543. 后序返回"向下最长链"，路径长度在合并处更新。O(n) / O(h)。"""
    best = 0

    def depth(node):
        nonlocal best
        if not node:
            return 0
        l, r = depth(node.left), depth(node.right)
        best = max(best, l + r)
        return 1 + max(l, r)

    depth(root)
    return best


def max_path_sum(root: Optional[TreeNode]) -> int:
    """LC 124. 与 543 同一模板：负贡献截断为 0。O(n) / O(h)。"""
    best = float("-inf")

    def gain(node):
        nonlocal best
        if not node:
            return 0
        l = max(gain(node.left), 0)
        r = max(gain(node.right), 0)
        best = max(best, node.val + l + r)
        return node.val + max(l, r)

    gain(root)
    return best


def lowest_common_ancestor(root: TreeNode, p: TreeNode, q: TreeNode) -> Optional[TreeNode]:
    """LC 236. 后序：左右各返回找到的节点；两边都有则当前是 LCA。O(n) / O(h)。"""
    if not root or root is p or root is q:
        return root
    l = lowest_common_ancestor(root.left, p, q)
    r = lowest_common_ancestor(root.right, p, q)
    if l and r:
        return root
    return l or r


def build_tree(preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
    """LC 105. 前序第一个是根，在中序里定位切分左右；哈希 O(1) 查位置。O(n) / O(n)。"""
    pos = {v: i for i, v in enumerate(inorder)}
    pre_idx = [0]

    def rec(lo: int, hi: int) -> Optional[TreeNode]:   # 中序区间 [lo, hi)
        if lo >= hi:
            return None
        val = preorder[pre_idx[0]]
        pre_idx[0] += 1
        node = TreeNode(val)
        mid = pos[val]
        node.left = rec(lo, mid)                        # 先建左，前序指针自然先消耗左子树
        node.right = rec(mid + 1, hi)
        return node

    return rec(0, len(inorder))


def is_valid_bst(root: Optional[TreeNode]) -> bool:
    """LC 98. 带上下界递归（不能只比较父子）。O(n) / O(h)。"""
    def check(node, lo, hi):
        if not node:
            return True
        if not (lo < node.val < hi):
            return False
        return check(node.left, lo, node.val) and check(node.right, node.val, hi)
    return check(root, float("-inf"), float("inf"))


def kth_smallest(root: Optional[TreeNode], k: int) -> int:
    """LC 230. 迭代中序数到第 k 个即停。O(h + k)。"""
    stack, cur = [], root
    while True:
        while cur:
            stack.append(cur)
            cur = cur.left
        cur = stack.pop()
        k -= 1
        if k == 0:
            return cur.val
        cur = cur.right


def serialize(root: Optional[TreeNode]) -> str:
    """LC 297. 前序 + '#' 占空位。"""
    out: List[str] = []

    def rec(node):
        if not node:
            out.append("#")
            return
        out.append(str(node.val))
        rec(node.left)
        rec(node.right)

    rec(root)
    return ",".join(out)


def deserialize(data: str) -> Optional[TreeNode]:
    """LC 297. 用迭代器按前序还原。"""
    it = iter(data.split(","))

    def rec():
        tok = next(it)
        if tok == "#":
            return None
        node = TreeNode(int(tok))
        node.left = rec()
        node.right = rec()
        return node

    return rec()


def path_sum_iii(root: Optional[TreeNode], target: int) -> int:
    """LC 437. 树上前缀和 + 哈希（LC 560 的树版），回溯时撤销计数。O(n) / O(h)。"""
    count = defaultdict(int)
    count[0] = 1
    ans = 0

    def dfs(node, pre):
        nonlocal ans
        if not node:
            return
        pre += node.val
        ans += count[pre - target]
        count[pre] += 1
        dfs(node.left, pre)
        dfs(node.right, pre)
        count[pre] -= 1                  # 离开这条路径时撤销

    dfs(root, 0)
    return ans


def flatten(root: Optional[TreeNode]) -> None:
    """LC 114. 原地展开为前序链表：把左子树接到右边、原右子树接到左子树最右。O(n) / O(1)。"""
    cur = root
    while cur:
        if cur.left:
            pre = cur.left
            while pre.right:
                pre = pre.right
            pre.right = cur.right
            cur.right, cur.left = cur.left, None
        cur = cur.right


class Tests(unittest.TestCase):
    def setUp(self):
        self.t = build([3, 9, 20, None, None, 15, 7])

    def test_traversals(self):
        self.assertEqual(inorder_iterative(self.t), [9, 3, 15, 20, 7])
        self.assertEqual(preorder_iterative(self.t), [3, 9, 20, 15, 7])
        self.assertEqual(level_order(self.t), [[3], [9, 20], [15, 7]])
        self.assertEqual(right_side_view(build([1, 2, 3, None, 5, None, 4])), [1, 3, 4])

    def test_basic(self):
        self.assertEqual(max_depth(self.t), 3)
        self.assertEqual(level_order(invert_tree(build([4, 2, 7, 1, 3, 6, 9]))), [[4], [7, 2], [9, 6, 3, 1]])
        self.assertTrue(is_symmetric(build([1, 2, 2, 3, 4, 4, 3])))
        self.assertFalse(is_symmetric(build([1, 2, 2, None, 3, None, 3])))

    def test_paths(self):
        self.assertEqual(diameter_of_binary_tree(build([1, 2, 3, 4, 5])), 3)
        self.assertEqual(max_path_sum(build([1, 2, 3])), 6)
        self.assertEqual(max_path_sum(build([-10, 9, 20, None, None, 15, 7])), 42)
        self.assertEqual(max_path_sum(build([-3])), -3)

    def test_lca(self):
        t = build([3, 5, 1, 6, 2, 0, 8, None, None, 7, 4])
        p, q = t.left, t.right                    # 5, 1
        self.assertIs(lowest_common_ancestor(t, p, q), t)
        self.assertIs(lowest_common_ancestor(t, p, t.left.right.right), p)   # 5 与 4 -> 5

    def test_build(self):
        t = build_tree([3, 9, 20, 15, 7], [9, 3, 15, 20, 7])
        self.assertEqual(level_order(t), [[3], [9, 20], [15, 7]])

    def test_bst(self):
        self.assertTrue(is_valid_bst(build([2, 1, 3])))
        self.assertFalse(is_valid_bst(build([5, 1, 4, None, None, 3, 6])))
        self.assertFalse(is_valid_bst(build([5, 4, 6, None, None, 3, 7])))    # 3 在右子树里但小于 5
        self.assertEqual(kth_smallest(build([5, 3, 6, 2, 4, None, None, 1]), 3), 3)

    def test_serialize(self):
        s = serialize(self.t)
        self.assertEqual(s, "3,9,#,#,20,15,#,#,7,#,#")
        self.assertEqual(level_order(deserialize(s)), [[3], [9, 20], [15, 7]])
        self.assertIsNone(deserialize(serialize(None)))

    def test_path_sum_iii(self):
        self.assertEqual(path_sum_iii(build([10, 5, -3, 3, 2, None, 11, 3, -2, None, 1]), 8), 3)

    def test_flatten(self):
        t = build([1, 2, 5, 3, 4, None, 6])
        flatten(t)
        vals, cur = [], t
        while cur:
            self.assertIsNone(cur.left)
            vals.append(cur.val)
            cur = cur.right
        self.assertEqual(vals, [1, 2, 3, 4, 5, 6])


if __name__ == "__main__":
    unittest.main(verbosity=1)

File: coding-interview/java/P05BinaryTree.java
import java.util.*;

/** 面试手撕代码（05）：二叉树。运行：java -ea P05BinaryTree */
public class P05BinaryTree {

    static class TreeNode {
        int val; TreeNode left, right;
        TreeNode(int v) { val = v; }
    }

    /** LeetCode 风格层序数组建树，null 为空。 */
    static TreeNode build(Integer... vals) {
        if (vals.length == 0 || vals[0] == null) return null;
        TreeNode root = new TreeNode(vals[0]);
        Deque<TreeNode> q = new ArrayDeque<>(List.of(root));
        int i = 1;
        while (!q.isEmpty() && i < vals.length) {
            TreeNode n = q.poll();
            if (i < vals.length && vals[i] != null) { n.left = new TreeNode(vals[i]); q.add(n.left); }
            i++;
            if (i < vals.length && vals[i] != null) { n.right = new TreeNode(vals[i]); q.add(n.right); }
            i++;
        }
        return root;
    }

    /** LC 94. 迭代中序。 */
    static List<Integer> inorderIterative(TreeNode root) {
        List<Integer> out = new ArrayList<>();
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode cur = root;
        while (cur != null || !stack.isEmpty()) {
            while (cur != null) { stack.push(cur); cur = cur.left; }
            cur = stack.pop();
            out.add(cur.val);
            cur = cur.right;
        }
        return out;
    }

    /** LC 102. 按层 BFS。 */
    static List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> out = new ArrayList<>();
        if (root == null) return out;
        Deque<TreeNode> q = new ArrayDeque<>(List.of(root));
        while (!q.isEmpty()) {
            List<Integer> level = new ArrayList<>();
            for (int n = q.size(); n > 0; n--) {
                TreeNode node = q.poll();
                level.add(node.val);
                if (node.left != null) q.add(node.left);
                if (node.right != null) q.add(node.right);
            }
            out.add(level);
        }
        return out;
    }

    /** LC 104. */
    static int maxDepth(TreeNode r) { return r == null ? 0 : 1 + Math.max(maxDepth(r.left), maxDepth(r.right)); }

    /** LC 226. */
    static TreeNode invertTree(TreeNode r) {
        if (r == null) return null;
        TreeNode l = invertTree(r.left);
        r.left = invertTree(r.right);
        r.right = l;
        return r;
    }

    /** LC 101. */
    static boolean isSymmetric(TreeNode r) { return r == null || mirror(r.left, r.right); }
    private static boolean mirror(TreeNode a, TreeNode b) {
        if (a == null || b == null) return a == b;
        return a.val == b.val && mirror(a.left, b.right) && mirror(a.right, b.left);
    }

    /** LC 543. 后序返回向下最长链，路径在合并处更新。 */
    static int diameter;
    static int diameterOfBinaryTree(TreeNode root) { diameter = 0; depth(root); return diameter; }
    private static int depth(TreeNode n) {
        if (n == null) return 0;
        int l = depth(n.left), r = depth(n.right);
        diameter = Math.max(diameter, l + r);
        return 1 + Math.max(l, r);
    }

    /** LC 124. 同一模板，负贡献截断为 0。 */
    static int best;
    static int maxPathSum(TreeNode root) { best = Integer.MIN_VALUE; gain(root); return best; }
    private static int gain(TreeNode n) {
        if (n == null) return 0;
        int l = Math.max(gain(n.left), 0), r = Math.max(gain(n.right), 0);
        best = Math.max(best, n.val + l + r);
        return n.val + Math.max(l, r);
    }

    /** LC 236. */
    static TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) return root;
        TreeNode l = lowestCommonAncestor(root.left, p, q), r = lowestCommonAncestor(root.right, p, q);
        if (l != null && r != null) return root;
        return l != null ? l : r;
    }

    /** LC 105. 前序 + 中序建树。 */
    static int preIdx;
    static TreeNode buildTree(int[] preorder, int[] inorder) {
        Map<Integer, Integer> pos = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) pos.put(inorder[i], i);
        preIdx = 0;
        return rec(preorder, pos, 0, inorder.length);
    }
    private static TreeNode rec(int[] pre, Map<Integer, Integer> pos, int lo, int hi) {
        if (lo >= hi) return null;
        TreeNode n = new TreeNode(pre[preIdx++]);
        int mid = pos.get(n.val);
        n.left = rec(pre, pos, lo, mid);
        n.right = rec(pre, pos, mid + 1, hi);
        return n;
    }

    /** LC 98. 上下界用 Long 避开 Integer 极值。 */
    static boolean isValidBST(TreeNode r) { return check(r, Long.MIN_VALUE, Long.MAX_VALUE); }
    private static boolean check(TreeNode n, long lo, long hi) {
        if (n == null) return true;
        if (n.val <= lo || n.val >= hi) return false;
        return check(n.left, lo, n.val) && check(n.right, n.val, hi);
    }

    /** LC 230. */
    static int kthSmallest(TreeNode root, int k) {
        Deque<TreeNode> stack = new ArrayDeque<>();
        TreeNode cur = root;
        while (true) {
            while (cur != null) { stack.push(cur); cur = cur.left; }
            cur = stack.pop();
            if (--k == 0) return cur.val;
            cur = cur.right;
        }
    }

    /** LC 297. 前序 + '#'。 */
    static String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        ser(root, sb);
        return sb.substring(0, sb.length() - 1);
    }
    private static void ser(TreeNode n, StringBuilder sb) {
        if (n == null) { sb.append("#,"); return; }
        sb.append(n.val).append(',');
        ser(n.left, sb); ser(n.right, sb);
    }
    static TreeNode deserialize(String data) {
        return des(new ArrayDeque<>(Arrays.asList(data.split(","))));
    }
    private static TreeNode des(Deque<String> toks) {
        String t = toks.poll();
        if (t.equals("#")) return null;
        TreeNode n = new TreeNode(Integer.parseInt(t));
        n.left = des(toks); n.right = des(toks);
        return n;
    }

    /** LC 437. 树上前缀和 + 哈希，回溯撤销。 */
    static int pathSumIII(TreeNode root, int target) {
        Map<Long, Integer> count = new HashMap<>();
        count.put(0L, 1);
        return dfs(root, 0L, target, count);
    }
    private static int dfs(TreeNode n, long pre, int target, Map<Long, Integer> count) {
        if (n == null) return 0;
        pre += n.val;
        int ans = count.getOrDefault(pre - target, 0);
        count.merge(pre, 1, Integer::sum);
        ans += dfs(n.left, pre, target, count) + dfs(n.right, pre, target, count);
        count.merge(pre, -1, Integer::sum);
        return ans;
    }

    /** LC 114. 原地展开。 */
    static void flatten(TreeNode root) {
        for (TreeNode cur = root; cur != null; cur = cur.right) {
            if (cur.left != null) {
                TreeNode pre = cur.left;
                while (pre.right != null) pre = pre.right;
                pre.right = cur.right;
                cur.right = cur.left;
                cur.left = null;
            }
        }
    }

    public static void main(String[] args) {
        TreeNode t = build(3, 9, 20, null, null, 15, 7);
        assert inorderIterative(t).equals(List.of(9, 3, 15, 20, 7));
        assert levelOrder(t).equals(List.of(List.of(3), List.of(9, 20), List.of(15, 7)));
        assert maxDepth(t) == 3;
        assert isSymmetric(build(1, 2, 2, 3, 4, 4, 3));
        assert levelOrder(invertTree(build(4, 2, 7, 1, 3, 6, 9))).get(1).equals(List.of(7, 2));
        assert diameterOfBinaryTree(build(1, 2, 3, 4, 5)) == 3;
        assert maxPathSum(build(-10, 9, 20, null, null, 15, 7)) == 42;
        TreeNode lca = build(3, 5, 1, 6, 2, 0, 8, null, null, 7, 4);
        assert lowestCommonAncestor(lca, lca.left, lca.right) == lca;
        assert lowestCommonAncestor(lca, lca.left, lca.left.right.right) == lca.left;
        assert levelOrder(buildTree(new int[]{3, 9, 20, 15, 7}, new int[]{9, 3, 15, 20, 7})).equals(levelOrder(t));
        assert isValidBST(build(2, 1, 3)) && !isValidBST(build(5, 4, 6, null, null, 3, 7));
        assert isValidBST(build(Integer.MAX_VALUE));
        assert kthSmallest(build(5, 3, 6, 2, 4, null, null, 1), 3) == 3;
        assert serialize(t).equals("3,9,#,#,20,15,#,#,7,#,#");
        assert levelOrder(deserialize(serialize(t))).equals(levelOrder(t));
        assert pathSumIII(build(10, 5, -3, 3, 2, null, 11, 3, -2, null, 1), 8) == 3;
        TreeNode f = build(1, 2, 5, 3, 4, null, 6);
        flatten(f);
        List<Integer> vals = new ArrayList<>();
        for (TreeNode c = f; c != null; c = c.right) { assert c.left == null; vals.add(c.val); }
        assert vals.equals(List.of(1, 2, 3, 4, 5, 6));
        System.out.println("P05BinaryTree OK");
    }
}
